### PR TITLE
Gateway: forward X-Fireworks-Genie header from client

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, there was no way to enable the "tracing" feature from Fireworks https://readme.fireworks.ai/docs/enabling-tracing This PR solves the problem by forwarding the `X-Fireworks-Genie` HTTP header to Fireworks if this HTTP header is set by the Gateway client.

Fixes CODY-2555

## Test plan

N/A

### CODE REVIEW

File: cmd/cody-gateway/internal/httpapi/completions/upstream.go [hunk 1]
```diff
@@ -302,7 +303,7 @@ func makeUpstreamHandler[ReqT UpstreamRequest](
 		}
 
 		// Run the request transformer.
-		methods.transformRequest(req)
+		methods.transformRequest(r, req)
```
Comment (eseliger) [inline cmd/cody-gateway/internal/httpapi/completions/upstream.go:306]: Might be nice to adopt the new naming scheme here as well (upstream and downstream req)

**Review by eseliger — APPROVED** [review]
As talked about in person, this is free of charge, and a client side change will need to explicitly enable this for requests. Seems low risk 👍

File: cmd/cody-gateway/internal/httpapi/completions/upstream.go (hunk 1 above)
Comment (olafurpg) [reply cmd/cody-gateway/internal/httpapi/completions/upstream.go]: Good idea. Done

**Comment by olafurpg** [comment]
FYI @rafax @chrsmith

**Comment by chrsmith** [comment]
Post-hoc LGTM. This changes looks good. Although I wish we had better names than "upstream" and "downstream" for these concepts. I personally find it kinda confusing. (Thinking that the caller would be "upstream", and the thing I was forwarding the call to would be "downstream".)

`incommingRequest` and `outboundRequest`? `callToCG` and `callTo3rdParty`? #namingishard

**Comment by olafurpg** [comment]
@chrsmith agreed on naming. I went with upstream/downstream because I saw several mentions of "upstream" referring to the request going to the LLM provider. I didn't want to introduce a third naming convention. No objections to doing a batch rename to something clearer

**Pull request merged**